go/libraries/doltcore/sqle/dtables/workspace.go

```diff
@@ -27,56 +27,368 @@ import (
+	fromRow := row[3+schemaLen:]
+
+	// If to Row has any non-nil values, then we need to do an update. Otherwise, insert.
+	wasDelete := true
+	for _, val := range toRow {
+		if val != nil {
+			wasDelete = false
+			break
+		}
+	}
+	wasInsert := true
+	for _, val := range fromRow {
+		if val != nil {
+			wasInsert = false
+			break
+		}
+	}
+
+	tableWriter := (*wtd.tableWriter)
+	if tableWriter == nil {
+		return fmt.Errorf("Runtime error: table writer is nil")
+	}
+
+	if wasInsert {
+		return tableWriter.Delete(c, toRow) // delete newly added row.
+	} else if wasDelete {
+		return tableWriter.Insert(c, fromRow) // restore deleted row.
+	} else {
+		return tableWriter.Update(c, toRow, fromRow) // restore updated row.
+	}
+}
+
+func (wtd *WorkspaceTableDeleter) Close(c *sql.Context) error {
+	// Resources released in StatementComplete
+	return nil
+}
+
+func (wtm *WorkspaceTableModifier) getWorkspaceTableWriter(ctx *sql.Context, targetStaging bool) (dsess.WriteSession, dsess.TableWriter, error) {
+	ds := dsess.DSessFromSess(ctx.Session)
+
+	setter := ds.SetWorkingRoot
+	if targetStaging {
+		setter = ds.SetStagingRoot
+	}
+
+	gst, err := dsess.NewAutoIncrementTracker(ctx, "dolt", wtm.ws)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	writeSession := writer.NewWriteSession(types.Format_DOLT, wtm.ws, gst, editor.Options{TargetStaging: targetStaging})
+
+	tableWriter, err := writeSession.GetTableWriter(ctx, doltdb.TableName{Name: wtm.tableName}, ctx.GetCurrentDatabase(), setter, targetStaging)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return writeSession, tableWriter, nil
+}
+
+// isTrue returns true if the value is a boolean true, or an int8 value of != 0. Otherwise, it returns false.
+func isTrue(value interface{}) bool {
+	switch v := value.(type) {
+	case bool:
+		return v
+	case int8:
+		return v != 0
+	default:
+		return false
+	}
+}
+
+// validateWorkspaceUpdate returns true IFF old and new row are identical - except the "staged" column. Updating that
+// column to TRUE or FALSE is the only update allowed, and any other update will result in 'valid' being false. If
+// valid is true, then 'staged' will be the value in the "staged" column of the new row.
+func validateWorkspaceUpdate(old, new sql.Row) (valid, staged bool) {
+	if len(old) != len(new) {
+		return false, false
+	}
+
+	isStaged := false
+
+	// Verify there are no changes in the columns other than the "staged" column.
+	for i := range new {
+		if i == stagedColumnIdx {
+			isStaged = isTrue(new[stagedColumnIdx])
+			// skip the "staged" column.
+			continue
+		}
+
+		if old[i] != new[i] {
+			return false, false
+		}
+	}
+	return true, isStaged
+}
+
+func (wt *WorkspaceTable) Deleter(_ *sql.Context) sql.RowDeleter {
+	cols := wt.headSchema.GetAllCols().Size()
+	modifier := WorkspaceTableModifier{
+		tableName: wt.userTblName,
+		headSch:   wt.headSchema,
+		schemaLen: cols,
+		ws:        wt.ws,
+		head:      wt.head,
+	}
+
+	return &WorkspaceTableDeleter{
```

Comment:
(minor) You could delete a little bit of code if you used WorkspaceTableModifier directly, instead of the specialized types WorkspaceTableDeleter and WorkspaceTableUpdater. Not a huge difference, but some duplicated methods could be removed, such as `StatementBegin`, `StatementComplete`, `Close`, `DiscardChanges`, etc. Alternatively, you could push those duplicated methods down into `WorkspaceTableModifier` and avoid having to duplicate them on the specialized types.